fix: select option does not delete in onBeforeMount

zhaozhongyu opened the pull request:
Please make sure these boxes are checked before submitting your PR, thank you!

* [x] Make sure you follow Element's contributing guide [English](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.en-US.md) | ([中文](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.zh-CN.md) | [Español](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.es.md) | [Français](https://github.com/element-plus/element-plus/blob/master/.github/CONTRIBUTING.fr-FR.md)).
* [x] Make sure you are merging your commits to `dev` branch.
* [x] Add some descriptions and refer to relative issues for your PR.
fix bug https://github.com/element-plus/element-plus/issues/2563, select option does not delete in onBeforeMount

Commit:
fix: select option does not delete in onBeforeMount

YunYouJun committed:
refactor(button): migrate css var (#2582)

YunYouJun committed:
refactor(switch): migrate css var (#2584)

* refactor(switch): migrate css var

* refactor(switch): remove switch var in common

YunYouJun committed:
refactor(skeleton): migrate css var (#2586)

YunYouJun committed:
refactor(pagination): migrate css var (#2587)

YunYouJun committed:
refactor(table+calendar): migrate css var (#2588)

* refactor(table): migrate css var

* refactor(calendar): migrate css var

zouhangwithsweet committed:
feat: new hook: `useCssVar` for inject custom css vars (#2547)

* feat: new hook: `usecssvars` for inject custom css vars

* perf: optimize usecssvar logic

* test: test css var

YunYouJun committed:
refactor(color): migrate color css var & simplify color box docs (#2589)

* refactor(color): migrate color css var & simplify color box docs

* Update unit-test.yml

Co-authored-by: jeremywu <15975785+JeremyWuuuuu@users.noreply.github.com>

YunYouJun committed:
refactor(tag): migrate css var (#2592)

Zenquan committed:
feat(autocomplete): optimize some const of autocomplete (#2593)

* feat(autocomplete): optimize some const of autocomplete

* feat(autocomplete): optimize some const of autocomplete

YunYouJun committed:
refactor(avatar): migrate css var (#2601)

kooriookami committed:
perf(affix): expose update function (#2603)

re #2594

jw-foss committed:
chore: Update README.md (#2608)

kooriookami committed:
docs(scrollbar): expose update function (#2607)

anguiao committed:
fix(style): ensure that the loading icon is inline (#2550)

JeremyWuuuuu commented on packages/select/src/option.vue:
```diff
@@ -67,20 +67,21 @@ export default defineComponent({
     } = toRefs(states)
 
     const vm = getCurrentInstance().proxy
+    const key = (vm as unknown as SelectOptionProxy).value
```
The proxy is essentially the instance of `Option` which owned this `value` but `Option` gets the `value` via its parent which is `props.value` rewriting this doesn't make any change on this I assume?

JeremyWuuuuu replied:
Would you elaborate the reason why were you doing this?

zhaozhongyu replied:
当我们手动输入一个值的时候, 创建option实例时使用了`vm.value`值作为`state.options`和`state.cachedOptions`的key, 此时可能只是输入了一个字符
![image](https://user-images.githubusercontent.com/8148957/126601136-621d351f-3ac2-456e-9572-1b20f84bcad3.png)

当继续输入更多字符后再去触发销毁事件时, 以前的代码里面尝试使用当前的`props.value`进行删除, 此时`state.options`和`state.cachedOptions`找不到对应的key, 导致内存泄漏和按方向键选中时会有隐藏options的问题
![image](https://user-images.githubusercontent.com/8148957/126601504-7da08c3d-698b-49a6-a986-f2bfd5d7c94a.png)

jw-foss committed:
chore: Update README.md (#2609)

* Update README.md

* Update README.md

0song committed:
chore: Update readme sponsor (#2616)

kooriookami committed:
perf(date-picker): shortcuts support function type (#2617)

re #2598

0song committed:
chore: update website (#2620)

0song committed:
chore: update readme (#2619)

jw-foss committed:
build(project): update bundle strategy

- Update the build script for generating *.d.ts files
- Update package.json module entry.
- Update CI pipeline node version
- Reorganized main bundle structure
- Update i18n functionalities

jw-foss committed:
update ci node version

msidolphin committed:
style(select-v2): optimized the style (#2628)

msidolphin committed:
test(select-v2): add some test cases & fix selectLabel/multiple-limit issue (#2626)

YunYouJun committed:
refactor(link): migrate css var (#2622)

0song committed:
chore: update changelog 1.0.2-beta.56 (#2634)

jw-foss committed:
fix(locale): Fix locale translation method (#2645)

- Fix defaultTranslation method that causes error

YunYouJun committed:
refactor(result): migrate css var (#2637)

YunYouJun committed:
refactor(carousel): migrate css var (#2638)

YunYouJun committed:
refactor(transfer): migrate css var (#2639)

YunYouJun committed:
refactor(dialog): migrate css var (#2641)

YunYouJun committed:
refactor(menu): migrate css var (#2642)

YunYouJun committed:
refactor(descriptions): migrate css var (#2643)

YunYouJun committed:
refactor(empty): migrate css var (#2644)

msidolphin committed:
fix: missing install & locale issue (#2647)

CoderMonkie committed:
fix(select): should not stop at invisible options by up and down (#2585)

* fix(select): should not stop at invisible options by up and down

this also fix another problem (#2562): about unselect user-created tag in ElSelect

fix #2563 #2562

* fix(select): checkDefaultFirstOption: exlucde disabled options

fix #2562 #2563

* test(select): add test for "default-first-option" (with navigation)

check default first option

re #2585

jw-foss committed:
fix(locale): fix default translate method

- Fix the wrong logic of default translate method

nabaonan committed:
fix: ondemand style lost (#2648)

msidolphin committed:
fix(select-v2): adjust the trigger conditions of focus and blur (#2650)

YunYouJun committed:
refactor(color): migrate all color-text var & simplify color docs (#2656)

YunYouJun committed:
refactor(collapse): migrate css var (#2661)

Jimmyzm committed:
style: remove bg-color of menu items at ':focus' state in vertical mode (#2659)

jw-foss committed:
feat(locale): locale-provider (#2662)

- Add a component for providing configs
- Add `use-locale` hooks into hooks for composable usage
- Add `use-locale` for `config-provider` to allow user to provide locale config via `react-like` API
- Add test cases against these new features

YunYouJun committed:
refactor(border): migrate css var & simplify docs (#2663)

wen-haoming committed:
fix(input): input maxlength calculate the correct length (#2654)

* fix(input): input maxlength calculate the correct length

* docs(input): Update maxlength comment

* fix(input): format code and remove upperLimit calculated attributes

* fix(input): props.maxlength is converted to number type

* docs(input): modify maxlength type

jw-foss committed:
refactor: i18n implementation (#2670)

jw-foss committed:
refactor(locale): i18n migration website code (#2671)

- Refactor website code accordingly.

iamkun committed:
chore: update i18n demo (#2672)

jw-foss committed:
chore(project): releace 1.0.2-beta.59 (#2674)

* chore(project): releace 1.0.2-beta.59

- Update change log

* update breaking change

jw-foss committed:
fix(locale): dayjs related i18n issue (#2678)

- Fix dayjs related i18n issue.

0song committed:
fix: add ElConfigProvider css file (#2679)

jw-foss committed:
fix(build): fix build script (#2680)

- Replace cp command with rsync

jw-foss committed:
fix(build): fix build script (#2683)

* fix(build): fix build script

- Replace cp command with rsync

* update build script

YunYouJun committed:
refactor(popover): migrate css var (#2687)

YunYouJun committed:
refactor(tooltip): migrate css var (#2686)

YunYouJun committed:
refactor(color): rename color-text to text-color (#2691)

* refactor(color): rename color-text to text-color

* fix(color): text-color typo

YunYouJun committed:
refactor(scrollbar): migrate css var (#2698)

YunYouJun committed:
fix(avatar): css property line-height (#2702)

YunYouJun committed:
refactor: migrate tree/popup/form/timeline-item css var (#2697)

mistery-li committed:
fix(message): fixed message iconClass not working (#2695)

"fix #2692"

Co-authored-by: tujin_li <tujin.li@gshopper.com>

msidolphin committed:
fix(notification): improve type definition and define closeAll function (#2705)

fix #2696

msidolphin committed:
docs(select-v2): add description of the height property (#2704)

fix #2700

kooriookami committed:
update i18n version (#2690)

wenfangdu committed:
fix(radio-button): `checked` should sync with state (#2699)

closes #2694

YunYouJun committed:
refactor(date-picker): migrate css var (#2689)

Co-authored-by: jeremywu <15975785+JeremyWuuuuu@users.noreply.github.com>

0song committed:
chore: update changelog 1.0.2-beta.63 (#2713)

YunYouJun committed:
refactor(var): simplify checkbox var & move transition to root (#2729)

YunYouJun committed:
fix(var): map merge default var (#2727)

msidolphin committed:
fix(scrollbar): the problem of ScrollbarDirKey being deleted (#2722)

msidolphin committed:
fix(input): no trigger input event when clear value (#2723)

fix #2688

GaliMu committed:
fix(message): using iconClass causes "el-message__icon" to be lost (#2709)

Co-authored-by: 穆建华 <galimu@foxmail.com>

YunYouJun committed:
refactor(backtop): migrate css var (#2711)

PxlBuzzard committed:
[types.ts] Fixed Nullable type for RefElement (#2730)

* Fixed Nullable type for RefElement

* Changed undefined to null

jw-foss committed:
fix(locale): inject locale for modal like components (#2737)

- Add a runtime locale object cache for modal components to use

jw-foss committed:
release(project): release 1.0.2-beta.64 (#2733)

* release(project): release 1.0.2-beta.64

- Update changelog

* update change log

* update change log

BeADre committed:
fix(input-number): fix style (#2738)

zycoJamie committed:
fix(date-picker): date-picker focus end-time resets start-time (#2290)

Commit:
Merge branch 'dev' of https://github.com/iBelieveLove/element-plus into dev
